File: pa3/List.java
//List.java
//This file is where the List ADT resides in for pa3.
//Samuel Shin, sayshin@ucsc.edu, pa3
class List
{
	private class Node
        {
		//fields
		Object data;
		Node next;
		Node previous;
		
		//constructor
		Node (Object data)
		{
			this.data = data;
			next = null;
			previous = null;
		}	
		
		//toString(): overrides Object's toString() method
		public String toString()
		{
			return String.valueOf(data);
		}

		public boolean equals (Object x)
		{
			if (x instanceof Node)
			{
				Object o = x;
				Node n = (Node) o;
				if (this == n)
				{
					return true;
				}
				else return false;	
			}
			else return false;
		}
	}
	
	//fields for List
	private Node front;
	private Node back;
	private Node cursor;
	private int length;
	private int index;

	List()//constructor for the List adt
	{
		front = null;
		back = null;
		cursor = null;
		length = 0;
		index = -1;
	}	
	
	int length() //returns the number of elements in the list
	{
		return length;	
	}	

	int index()
	{
		if (cursor != null) //if cursor is defined, returns the index of the cursor element,
				    //otherwise returns -1.
		{	
			return index;
		}
		else 
		{
			return -1;
		}
	}

	Object front() //returns front element. Pre: length() > 0
	{
		if (length() > 0)
		{
			return front.data;
		}
		else return -1;	
	}

	Object back() //returns back element. Pre: length() > 0
	{
		if (length() > 0)
		{
			return back.data;
		}
		else return -1;
	}

	Object get() //returns cursor element. Pre: length() > 0, index >= 0
	{
		if (length() > 0 && index >= 0)
		{
			return cursor.data;
		}
		else return -1;
	}

	public boolean equals (Object x) //returns true if and only if this List and L are the same
				//integer sequence. The states of the cursors in the two lists
				//are not used in determining equality.
	{
		if (x instanceof List)
		{
			Object o = x;
			List L = (List) o;	
			moveFront();
			L.moveFront();
			if (length != L.length)
			{
				return false;
			}
			else 
			{
				while (cursor != null)
				{
					if (!(cursor.data.equals(L.cursor.data)))
					{
						return false;		
					}
					moveNext();
					L.moveNext();
				}
			}
			return true;
		}
		else return false;
		
	}

	void clear() //resets this list to its original empty state
	{
		length = 0;
		index = 0;
		front = null;
		back = null;
		cursor = null;
	}	

	void moveFront() //if list is non-empty, places the cursor under the front element
			 //otherwise does nothing
	{
		if (length() > 0)
		{
			cursor = front;
			index = 0;
		}
	}

	void moveBack() //if list is non-empty, places the cursor under the back element
			//otherwise does nothing
	{
		if (length() > 0)
		{
			cursor = back;
			index = length - 1;
		}
	}

	void movePrev() //if cursor is defined and not at front, moves cursor one step toward
			//front of this list, if cursor is defined and at front, cursor becomes
			//undefined, if cursor is undefined does nothing
	{
		if (cursor != null && cursor != front) 
		{
			cursor = cursor.previous;
			index--;
		}
		else if (cursor != null && cursor == front)
		{
			cursor = null;
			index = -1;
		}
		else if (cursor == null)
		{}
	}

	void moveNext() //if cursor is defined and not at back, moves cursor one step toward
			//back of this list, if cursor is defined and at back, cursor becomes
			//undefined, if cursor is undefiend does nothing
	{
		if (cursor != null && cursor != back)
		{
			cursor = cursor.next;
			index++;
		}
		else if (cursor != null && cursor == back)
		{
			cursor = null;
			index = -1;
		}
		else if (cursor == null)
		{}
	}

	void prepend (Object data) //insert new element into this List. if List is non-empty,
				//insertion takes place before front element.
	{
		if (length() > 0)
		{
			Node n = new Node(data);
			front.previous = n;
			n.next = front;
			front = n;
			length++;
		}
		else //if list is empty new element is the first element in list
		{
			Node n = new Node(data);
			front = n;
			back = n;
			length++;	
		}
		
		if (cursor != null)
		{
			index++;
		}
	}
	
	void append (Object data) //insert new element into this List. if List is non-empty,
			       //insertion takes place after back element
	{
		if (length() > 0)
		{
			Node n = new Node(data);
			back.next = n;
			n.previous = back;
			back = n;
			length++;
		}
		else //if list is empty new element is the first element in list
		{
			Node n = new Node(data);
			front = n;
			back = n;
			length++;
		}

		if (cursor == back)
		{
			index++;
		}
	}

	void insertBefore (Object data) //insert new element before cursor
				     //pre: length() > 0, index >= 0
	{
		if (length() > 0 && index >= 0)
		{
			if (cursor == front) //if the cursor is defined as front
			{
				prepend(data);
				
			}
			else //if cursor is somewhere between front(exclusive) and back(inclusive) of list add it in before cursor
			{
				Node n = new Node(data);
				cursor.previous.next = n;
				n.previous = cursor.previous;
				n.next = cursor;
				cursor.previous = n;
				index++;
				length++;
			}
		}
	}

	void insertAfter (Object data) //insert new element after cursor
				    //pre: length() > 0, index >= 0
	{
		if (length() > 0 && index >= 0)
		{
			if (cursor == back)
			{
				append(data);
			}
			else //if cursor is somewhere between front(inclusive) and back (exclusive) of list add it in after cursor
			{
				Node n = new Node(data);
				cursor.next.previous = n;
				n.next = cursor.next;
				n.previous = cursor;
				cursor.next = n;
				length++;
			}
		}
	}

	void deleteFront() //deletes the front element. Pre: length() > 0
	{
		if (length() > 0)
		{
			if (length() == 1)
			{
				clear();
			}
			else
			{
				front = front.next;
				front.previous = front.previous.previous;	
				length--;
			}
		}
		if (index > 0)//if the index isn't at the front
		{
			index--;
		}
		else if (index == 0)
		{
			cursor = null;
			index = -1;
		}
	}
	
	void deleteBack() //deletes the back element. Pre: length() > 0
	{
		if (length() > 0)
		{
			if (length() == 1)
			{
				clear();
			}
			else 
			{
				back = back.previous;
				back.next = back.next.next;
				length--;
			}
		}
		if (index == length())
		{
			cursor = null;
			index = -1;
		}
		
	}

	void delete() //deletes cursor element, making cursor undefined.
		      //Pre: length() > 0, index >= 0
	{
		if (length() == 1 && index >= 0)
		{
			clear();
		}
		else if (length() > 0 && index >= 0)
		{
			if (cursor == front)
			{
				deleteFront();
			}
			else if (cursor == back)
			{
				deleteBack();
			}
			else 
			{
				cursor.previous.next = cursor.next;
				cursor.next.previous = cursor.previous;
				cursor = null;
				index = -1;
				length--;
			}
		}
	}
	
	public String toString() //overrides object's toString method. Returns a string
				 //representation of this List consisting of a space
				 //separated sequence of integers, with front on left.
	{
		String str = "";
		moveFront();
		while (cursor != null)
		{
			if (cursor == back)
			{
				str += cursor.data + "";
				moveNext();
			}
			else 
			{
				str += cursor.data + " ";
				moveNext();
			}
		}
		return str;
	}

	/*List copy() //returns a new List representing the same integer sequence as this
		    //List. The cursor in the new list is undefined, regardless of the 
		    //state of the cursor in this List. This List is unchanged
	{
		List clone = new List();
		moveFront();
		while (cursor != null)
		{
			clone.append(cursor.data);
			moveNext();
		}
		moveFront();
		return clone;
	}*/	
}

File: pa4/backup/List.c
//List.c
//Implementation for the List ADT
//written by Samuel Shin, sayshin@ucsc.edu, pa2

#include "List.h"
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

//structs-------------------------------------------------------------------------------------------

//private NodeObj type
typedef struct NodeObj
{
	int data;
	struct NodeObj* next;
	struct NodeObj* previous;
}NodeObj;

//private Node type
typedef NodeObj* Node;

//private ListObj type
typedef struct ListObj
{
	int length;
	int index;
	Node front;
	Node back;
	Node cursor;
}ListObj;

//Constructors-Destructors--------------------------------------------------------------------------

//newNode()
//Returns a reference to a new Node object. Initializes data,next, and previous fields.
//private.
Node newNode(int data)
{
	Node N = malloc(sizeof(NodeObj));
	N->data = data;
	N->next = NULL;
	N->previous = NULL;
	return N;
}

//freeNode()
//Frees heap memory pointed to by *pN, sets *pN to NULL
//private

void freeNode(Node* pN)
{
	if (pN != NULL && *pN != NULL)
	{
		free(*pN);
		*pN = NULL;
	}
}

//NewList()
//returns a reference to new empty List object

List newList(void)
{
	List L;
	L = malloc(sizeof(ListObj));
	L->front = NULL;
	L->back = NULL;
	L->cursor = NULL;
	L->length = 0;
	L-> index = -1;
	return L;
}

//freeList()
//frees all heap memory associated with List *pL and sets *pL to NULL
void freeList (List* pL)
{
	if (pL != NULL && (*pL) != NULL)
	{
		clear(*pL);
		free(*pL);
		pL = NULL;
	}
		
}

//access functions------------------------------------------------------------------------------------------------------------------------------------------------------


int length(List L)
{
	if (L == NULL)
	{
		printf("List Error: calling length(List L) on NULL List reference");
		exit(1);
	}
	else 
		return L->length;
}

int index(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling index(List L) on NULL List reference");
                exit(1);
        }
	if (L->cursor != NULL)
	{
		return L->index;
	}
	else 
		return -1;
}

int front(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling front(List L) on NULL List reference");
                exit(1);
        }
	if (L->length > 0)
	{
		return L->front->data;
	}
	return -1;	
}

int back(List L)
{
        if (L == NULL)
        {
                printf("List Error: calling back(List L) on NULL List reference");
                exit(1);
        }
        if (L->length > 0)
        {
                return L->back->data;
        }
	return -1;
}

int get(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling get(List L) on NULL List reference");
                exit(1);
        }
	if (L->length > 0 && L->index >= 0)
	{
		return L->cursor->data;
	}
	else 
		return -1;
}


int equals(List A, List B)
{
	Node N, M;
	if (A == NULL || B == NULL)
        {
                printf("List Error: calling equals(List A, List B) on NULL List reference");
                exit(1);
        }
	if (A->length != B->length)
	{
		return 0;
	}
	else 
	{
		N = A->front;
		M = B->front;
		while (N != NULL || M != NULL)
		{
			if (N->data != M->data)
			{
	
			return 0;
			}
			N = N->next;
			M = M->next;
		}
		return 1;
	}
}

//Manipulation procedures---------------------------------------------------------------------------

void clear(List L)
{
	
	if (L == NULL)
        {
                printf("List Error: calling clear(List L) on NULL List reference");
                exit(1);
        }
	
	
	Node N = L->front;
	Node temp;
	while (N != NULL)
	{
		temp = N;
		N = N->next;
		freeNode(&temp);	
	}	

	L->length = 0;
	L->index = -1;
	L->front = NULL;		
	L->back = NULL;
	L->cursor = NULL;
	
	
}

void moveFront(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling moveFront(List L) on NULL List reference");
                exit(1);
        }
	if (L->length > 0)
	{
		L->cursor = L->front;
		L->index = 0;
	}
}

void moveBack(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling moveBack(List L) on NULL List reference");
                exit(1);
        }
	if (L->length > 0)
	{
		L->cursor = L->back;
		L->index = L->length - 1;
	}
}

void movePrev(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling movePrev(List L) on NULL List reference");
                exit(1);
        }
	if (L->cursor != NULL && L->cursor != L->front)
	{
		L->cursor = L->cursor->previous;
		L->index--;
	}
	else if (L->cursor != NULL && L->cursor == L->front)
	{
		L->cursor = NULL;
		L->index = -1;
	}
	else if (L->cursor == NULL)
	{}
}

void moveNext(List L)
{
        if (L == NULL)
        {
                printf("List Error: calling moveNext(List L) on NULL List reference");
                exit(1);
        }
        if (L->cursor != NULL && L->cursor != L->back)
        {
                L->cursor = L->cursor->next;
		L->index++;
        }
        else if (L->cursor != NULL && L->cursor == L->back)
        {
                L->cursor = NULL;
		L->index = -1;
        }
	else if (L->cursor == NULL)
	{}
}

void prepend(List L, int data)
{
	if (L == NULL)
        {
                printf("List Error: calling prepend(List L, int data) on NULL List reference");
                exit(1);
        }
	if (L->length > 0)
	{
		Node N = newNode(data);
		L->front->previous = N;
		N->next = L->front;
		L->front = N;
		L->length++;
	}
	else 
	{
		Node N = newNode(data);
		L->front = N;
		L->back = N;
		L->length++;
	}

	if (L->cursor != NULL)
	{
		L->index++;
	}
}

void append(List L, int data)
{
	if (L == NULL)
        {
                printf("List Error: calling append(List L, int data) on NULL List reference");
                exit(1);
        }	
	if (L->length > 0)
	{
		Node N = newNode(data);
		L->back->next = N;
		N->previous = L->back;
		L->back = N;
		L->length++;
	}
	else 
	{
		Node N = newNode(data);
		L->front = N;
		L->back = N;
		L->length++;
	}
	if (L->cursor == L->back)
	{
		L->index++;
	}
}

void insertBefore(List L, int data)
{
	if (L == NULL)
        {
                printf("List Error: calling insertBefore(List L, int data) on NULL List reference");
                exit(1);
        }
	if (L->length > 0 && L->index >= 0)
	{
		if (L->cursor == L->front)
		{
			prepend(L, data);
		}
		else 
		{
			Node N = newNode(data);
			L->cursor->previous->next = N;
			N->previous = L->cursor->previous;
			N->next = L->cursor;
			L->index++;
			L->length++;
		}
	}
}

void insertAfter(List L, int data)
{
	if (L == NULL)
        {
                printf("List Error: calling insertAfter(List L, int data) on NULL List reference");
                exit(1);
        }	
	if (L->length > 0 && L->index >= 0)
	{
		if (L->cursor == L->back)
		{
			append(L, data);
		}
		else 
		{
			Node N = newNode(data);
			L->cursor->next->previous = N;
			N->next = L->cursor->next;
			N->previous = L->cursor;
			L->cursor->next = N;
			L->length++;
		}
	}
}

void deleteFront(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling deleteFront(List L) on NULL List reference");
                exit(1);
        }
	if (L->length > 0)
	{
		if (L->length == 1)
		{
			clear(L);
			freeNode(&L->front);
		}
		else 
		{
			Node temp;
			temp = L->front;
			L->front = L->front->next;
			L->front->previous = L->front->previous->previous;
			freeNode(&temp);
			L->length--;
		}
	}
	if (L->index > 0)
	{
		L->index--;
	}
	else if (L->index == 0)
	{
		L->cursor = NULL;
		L->index = -1;
	}
}

void deleteBack(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling deleteBack(List L) on NULL List reference");
                exit(1);
        }	
	if (L->length > 0)
	{
		if (L->length == 1)
		{
			clear(L);
			freeNode(&L->back);
		}
		else 
		{
			Node temp;
			temp = L->back;
			L->back = L->back->previous;
			L->back->next = L->back->next->next;
			freeNode(&temp);
			L->length--;
		}
	}
	if (L->index == L->length)
	{
		L->cursor = NULL;
		L->index = -1;
	}
}

void delete(List L)
{
	if (L == NULL)

        {
                printf("List Error: calling delete(List L) on NULL List reference");
                exit(1);
        }
	if (L->length == 1 && L->index >= 0)
	{
		clear(L);
		freeNode(&L->front);
	}
	else if (L->length > 0 && L->index >= 0)
	{
		if (L->cursor == L->front)
		{
			deleteFront(L);
		}
		else if (L->cursor == L->back)
		{
			deleteBack(L);
		}
		else 
		{
			Node temp;
			temp = L->cursor;
			L->cursor->previous->next = L->cursor->next;
			L->cursor->next->previous = L->cursor->previous;
			freeNode(&temp);
			L->cursor = NULL;
			L->index = -1;
			L->length--;
		}
	}
}

void printList(FILE* out, List L)
{
	Node N;
	if (L == NULL)
        {
                printf("List Error: calling printList(List L) on NULL List reference");
                exit(1);
        }
	N = L->front;
	while (N != NULL)
	{
		if (N->next != NULL)
		{
			fprintf(out, "%d ", N->data);
			N = N->next;
		}
		else 
		{
			fprintf(out, "%d", N->data);
			N = N->next;
		}
	}
	freeNode(&N);
}

List copyList(List L)
{
	if (L == NULL)
        {
                printf("List Error: calling copyList(List L) on NULL List reference");
                exit(1);
        }
	List clone = newList();
	Node N = L->front;
	while (N != NULL)
	{
		append(clone, N->data);
		N = N->next;
	}
	freeNode(&N);
	return clone;
}
